### PR TITLE
little cleanup of providers in preparation for adding forms provider

### PR DESCRIPTION
**Pull request by boyvinall**
- moved each provider implementation into separate package
- make basic provider comply with the core interface

### CODE REVIEW

File: example/idp/main.go [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
-	"github.com/janekolszak/idp/providers/cookie"
-
 	"flag"
 	"fmt"
-	"github.com/gorilla/sessions"
-	"github.com/julienschmidt/httprouter"
 	"net/http"
 	"text/template"
 	"time"
+
```
Comment (janekolszak) [inline example/idp/main.go]: pointless, remove

File: providers/basic/basic-auth.go [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
 
 import (
-	"github.com/janekolszak/idp/core"
-	"github.com/janekolszak/idp/helpers"
-
 	"fmt"
-	"golang.org/x/crypto/bcrypt"
 	"net/http"
+
+	"github.com/janekolszak/idp/core"
```
Comment (janekolszak) [inline providers/basic/basic-auth.go]: pointless, remove

File: providers/basic/basic-auth.go [hunk 3]
```diff
@@ -33,23 +32,23 @@ func (c *BasicAuth) Check(r *http.Request) (user string, err error) {
 	user, pass, ok := r.BasicAuth()
 	if !ok {
 		err = core.ErrorAuthenticationFailure
-		return
+		return user, err
```
Comment (janekolszak) [inline providers/basic/basic-auth.go]: this is done automatically by golang

File: providers/cookie-auth.go [hunk 4]
```diff
@@ -1,109 +0,0 @@
-package providers
```
Comment (janekolszak) [inline providers/cookie-auth.go]: I don't think it's time for cleanup yet.

File: providers/basic/basic-auth.go (hunk 2 above)
Comment (boyvinall) [inline providers/basic/basic-auth.go]: Sorry, the tooling I use (goreturns) did it like this.  Not trying to annoy..  I like the way goreturns does this, but it's your project so happy to play by your rules, of course.

File: providers/basic/basic-auth.go [hunk 5]
```diff
@@ -32,23 +32,23 @@ func (c *BasicAuth) Check(r *http.Request) (user string, err error) {
...
 	}
 
 	hash, err := c.Htpasswd.Get(user)
 	if err != nil {
 		// Prevent timing attack
 		bcrypt.CompareHashAndPassword([]byte{}, []byte(pass))
 		err = core.ErrorAuthenticationFailure
-		return
+		return "", err
```
Comment (janekolszak) [inline providers/basic/basic-auth.go]: remove

File: providers/basic/basic-auth.go [hunk 6]
```diff
@@ -32,23 +32,23 @@ func (c *BasicAuth) Check(r *http.Request) (user string, err error) {
...
 	}
 
 	err = bcrypt.CompareHashAndPassword([]byte(hash), []byte(pass))
 	if err != nil {
 		err = core.ErrorAuthenticationFailure
 	}
 
-	return
+	return user, err
```
Comment (janekolszak) [inline providers/basic/basic-auth.go]: remove

**Comment by janekolszak** [comment]
I like the cleanup, thanks. If you're planning to implement the form provider let me know so we don't duplicate the work.

File: providers/basic/basic-auth.go (hunk 3 above)
Comment (boyvinall) [inline providers/basic/basic-auth.go]: Agree, but you can sometimes find that with a `:=` assignment you can end up defining a new variable that overrides the one from the outer context. Was doing it for clarity of that in this case.

**Comment by boyvinall** [comment]
Cool.  I'm always conscious of trying not to annoy at first .. definitely not trying to create work for you. I'll implement the forms on a feature branch for you to take a look over though.

**Pull request merged**